src/arizona_server.erl

```diff
@@ -32,49 +30,53 @@
 %% Macros
 -define(LISTENER, arizona_http_listener).
 -define(PERSIST_KEY, arizona_dispatch).
+-elvis([{elvis_style, no_macros, #{allow => ['LISTENER', 'PERSIST_KEY']}}]).
+
+-type opts() :: #{
+    scheme => http | https,
+    transport => ranch_tcp:opts(),
+    proto => #{
+        proto => cowboy:opts(),
+        host => '_' | iodata(), % as per cowboy_router:route_match(),
+        routes => list(),
+        live_reload => boolean()
+    }
+}.
+-export_type([opts/0]).
 
 %% --------------------------------------------------------------------
 %% API functions.
 %% --------------------------------------------------------------------
 
+-spec start() -> pid().
 start() ->
-    start(application:get_env(arizona, endpoint, #{})).
+    start(arizona_cfg:endpoint()).
 
-start(Opts) when is_map(Opts) ->
+-spec start(Opts) -> pid()
+    when Opts :: opts().
+start(Opts) ->
     do_start(norm_opts(Opts)).
 
+-spec route(Req) -> Routed
+    when Req :: cowboy_req:req(),
+         Routed :: {Req, Env},
+         Env :: cowboy_middleware:env().
 route(Req) ->
     #{path := Path} = cowboy_req:match_qs([path], Req),
-    cowboy_router:execute(Req#{path => Path},
-                          #{dispatch => {persistent_term, ?PERSIST_KEY}}).
+    {ok, Req, Env} = cowboy_router:execute(Req#{path => Path},
+                                           #{dispatch => {persistent_term, ?PERSIST_KEY}}),
+    {Req, Env}.
 
 %% --------------------------------------------------------------------
 %% Internal functions.
 %% --------------------------------------------------------------------
 
 do_start(#{scheme := http, transport := Transport, proto := Proto}) ->
     {ok, Pid} = cowboy:start_clear(?LISTENER, Transport, Proto),
-    print_running(http),
-    {ok, Pid};
+    Pid;
 do_start(#{scheme := https, transport := Transport, proto := Proto}) ->
     {ok, Pid} = cowboy:start_tls(?LISTENER, Transport, Proto),
-    print_running(https),
-    {ok, Pid}.
-
-print_running(Scheme) ->
-    Info = ranch:info(?LISTENER),
-    {ip, Ip} = proplists:lookup(ip, Info),
-    {port, Port} = proplists:lookup(port, Info),
-    ?LOG_INFO("Arizona is running at ~w://~s:~p~n",
-              [Scheme, ip_to_str(Ip), Port]).
-
-ip_to_str(Ip) ->
-    case inet:ntoa(Ip) of
-        {error, einval} ->
-            error({invalid_ip, Ip});
-        Str ->
-            Str
-    end.
+    Pid.
```

Comment:
Wouldn't it be nice to `?LOG_INFO` the URL? Printing it in the shell makes it easy to `CTRL+click` on the link to open the browser.

Reply:
I guess, if you want. I'm thinking about this in terms of "production environments". The less logging the better, but we can maybe add some metadata to it, so the logs are controllable from the consumer.

Reply:
> The less logging the better, but we can maybe add some metadata to it, so the logs are controllable from the consumer.

I agree. It's important to know "where" the app is starting, but I think this is more important in a development environment.